r/src/RecordBatch.cpp

```diff
@@ -94,6 +95,27 @@ int RecordBatch__to_file(const std::shared_ptr<arrow::RecordBatch>& batch,
   return offset;
 }
 
+// [[Rcpp::export]]
+RawVector RecordBatch__to_raw(const std::shared_ptr<arrow::RecordBatch>& batch) {
+  MemoryPool* pool = default_memory_pool();
+
+  std::shared_ptr<ResizableBuffer> buffer;
+  R_ERROR_NOT_OK(AllocateResizableBuffer(pool, 0, &buffer));
+
+  std::unique_ptr<io::BufferOutputStream> sink;
+  sink.reset(new io::BufferOutputStream(buffer));
+
+  std::shared_ptr<arrow::ipc::RecordBatchWriter> writer;
+  R_ERROR_NOT_OK(arrow::ipc::RecordBatchStreamWriter::Open(sink.get(), batch->schema(), &writer));
+
+  R_ERROR_NOT_OK(writer->WriteRecordBatch(*batch));
+  R_ERROR_NOT_OK(writer->Close());
+
+  RawVector res(buffer->size());
+  std::copy(buffer->data(), buffer->data() + buffer->size(), res.begin()) ;
```

Comment:
There are functions to compute the output size so you can allocate once and then have `RecordBatchStreamWriter` write into the preallocated memory

Reply:
For starters, you can use `arrow::ipc::WriteRecordBatchStream` and `MockOutputStream` to write the stream, and to determine the payload size, respectively.

Reply:
Thanks for the feedback, @wesm. I haven't found an easy way to get the output size, if you have at hand a few pointers, that would be great. But yes, I agree, would be much better to allocate the memory in R and then just dump the contents. I'm also looking into not even allocating the memory but rather, passing directly the stream into an R socket so I might not even need to implement the `to_raw()` interface, will keep this PR up to date. Thanks again for the feedback.

Reply:
Never mind, just read your previous comment, using `MockOutputStream` seems reasonable.

Reply:
Changed.